### DISCUSSION TITLE
Provide a way to define build time global variables like webpack define plugin

### DISCUSSION THREAD

**Discussion by shirshendubhowmick** [remix-run/remix#3932] [Proposals]
Is there any alternative / equivalent way of defining build time global variable during build like we can do with webpack define plugin ?

Use case:

I have to pass version information (which is only available during build time) to the application for usage with `Sentry.init`

**Comment by sergiodxa** [comment]
There’s no way to customize the compiler, you can run your own script before and after Remix build manually thought

**Reply by shirshendubhowmick** [reply]
Any plan of adding support for a pluggable / customisable build system ?

I have worked with multiple CSR apps bundled using webpack and from that experience I have seen certain use cases which were solved by developing custom loader / plugin.

A few examples will be:
* Custom CSS loading pipeline
* Automatic test ID generation and insertion
* Multi tenancy support

Porting of these apps to Remix will be tricky for the use cases mentioned above.

**Comment by michaelgmcd** [comment]
I am running into this as well (for Sentry too). I've ended up patching @remix-run/dev to allow custom "define" values in the remix config to be replaced in the browser build. Doing this in a custom script is doable but becomes cumbersome when running the dev server.

**Reply by gustavopch** [reply]
I've also patched. Would like to see a `define` config officially supported by Remix.

**Comment by shirshendubhowmick** [comment]
Have changed this from Q&A to proposal

**Comment by sahithyandev** [comment]
Any improvements on this proposal?

**Reply by gustavopch** [reply]
Pretty sure you can simply use Vite's `define` config these days.

**Reply by tombohub** [reply]
but remix doesn't have vite.config. version 2.6.0

**Reply by cliffordfajardo** [reply]
@tombohub 

Creating a new remix app on v2.6.0 & I see the `vite.config.ts` file?
```
npx create-remix --template remix-run/remix/templates/unstable-vite-express
```

![CleanShot 2024-02-14 at 06 24 27@2x](https://github.com/remix-run/remix/assets/6743796/ec236128-0c3d-4e04-96f8-5370c9235512)

**Reply by tombohub** [reply]
@cliffordfajardo thanks, and this template can also run on fly? Should I just use plain vite template?

**Reply by cliffordfajardo** [reply]
yes, I've gotten remix to work on fly
I remeber being surprised by how simple the process was, it was just a few commands to get it deployed. I used the remix express template

I think its pretty much something like this ([docs may have changed](https://fly.io/docs/js/frameworks/remix/)), you might have to fiddle with it. you can create another github discussion for remix+fly since this original discussion might get confusing 😉 
```
cd my-remix-app
fly launch
```
